keystore/keystore.go

```diff
@@ -0,0 +1,244 @@
+package keystore
+
+import (
+	"context"
+	"crypto/ed25519"
+	"encoding/json"
+	"fmt"
+	"sync"
+	"time"
+
+	"golang.org/x/crypto/curve25519"
+
+	gethkeystore "github.com/ethereum/go-ethereum/accounts/keystore"
+	gethcrypto "github.com/ethereum/go-ethereum/crypto"
+	"github.com/smartcontractkit/chainlink-common/keystore/internal"
+	"github.com/smartcontractkit/chainlink-common/keystore/serialization"
+	"github.com/smartcontractkit/chainlink-common/keystore/storage"
+	"google.golang.org/protobuf/proto"
+)
+
+type KeyType string
+
+const (
+	// Hybrid encryption (key exchange + encryption) key types.
+	// Naming schema is generally <key exchange algorithm><encryption algorithm>.
+	// Except for widely used/commonly paired encryption algorithms, we
+	// omit the encryption algorithm. So for example X25519 with ChaCha20Poly1305
+	// (via box) is specified just as X25519.
+	// X25519:
+	// - X25519 for ECDH key exchange.
+	// - Box for encryption (ChaCha20Poly1305)
+	X25519 KeyType = "X25519"
+	// TODO: EcdhP256:
+	// - ECDH on P-256
+	// - Encryption with AES-GCM.
+
+	// Digital signature key types.
+	// Ed25519:
+	// - Ed25519 for digital signatures.
+	Ed25519 KeyType = "ed25519"
+	// EcdsaSecp256k1:
+	// - ECDSA on secp256k1 for digital signatures.
+	EcdsaSecp256k1 KeyType = "ecdsa-secp256k1"
+)
+
+var AllKeyTypes = []KeyType{X25519, Ed25519, EcdsaSecp256k1}
+
+type ScryptParams struct {
+	N int
+	P int
+}
+
+var (
+	DefaultScryptParams = ScryptParams{
+		N: gethkeystore.StandardScryptN,
+		P: gethkeystore.StandardScryptP,
+	}
+	FastScryptParams ScryptParams = ScryptParams{
+		N: 1 << 14,
+		P: 1,
+	}
+)
+
+// KeyInfo is the information about a key in the keystore.
+// Public key may be empty for non-asymmetric key types.
+type KeyInfo struct {
+	Name      string
+	KeyType   KeyType
+	CreatedAt time.Time
+	PublicKey []byte
+	Metadata  []byte
+}
+
+// newKeyInfo is a private constructor that ensures all fields are set explicitly.
+func newKeyInfo(name string, keyType KeyType, createdAt time.Time, publicKey []byte, metadata []byte) KeyInfo {
+	return KeyInfo{
+		Name:      name,
+		KeyType:   keyType,
+		CreatedAt: createdAt,
+		PublicKey: publicKey,
+		Metadata:  metadata,
+	}
+}
+
+type Keystore interface {
+	Admin
+	Reader
+	Signer
+	Encryptor
+}
+
+type Unimplemented struct{}
+
+func (Unimplemented) mustEmbedUnimplementedKey() {}
```

Comment:
If we expect external implementations of these interfaces, then they should each have their own `UnimplmentedX`, with methods satisfying `X`. In the current form, adding a new method is still a breaking change.

Reply:
The point of an exported `Unimplemented` type is to be embedded in implementations so that we may add methods to the interface in the future without breaking compatibility. If we don't want other implementations at all, then we should drop this type - but that seems unnecessarily restrictive, why make it impossible to make a simple fake/mock implementation?

Reply:
ok misunderstood its purpose then. I was thinking of just preventing external implementations all together (even a fake/mock implementation would be supplied in this package). Can switch to the other model of allow external impls which would embed the type so they can be extended